controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1686,23 +1699,45 @@ private StreamSegmentRecord newSegmentRecord(int epoch, int segmentNumber, long
     // endregion
 
     // region transactions
-    abstract CompletableFuture<Version> createNewTransaction(final int epoch, final UUID txId, final byte[] record);
+    abstract CompletableFuture<Version> createNewTransaction(final UUID txId, final byte[] record);
 
-    abstract CompletableFuture<Data> getActiveTx(final int epoch, final UUID txId);
+    abstract CompletableFuture<Data> getActiveTx(final UUID txId);
 
-    abstract CompletableFuture<Version> updateActiveTx(final int epoch,
-                                                       final UUID txId,
-                                                       final Data data);
+    abstract CompletableFuture<Version> updateActiveTx(final UUID txId, final Data data);
+
+    /**
+     * This method is called for adding a transaction to committing order. 
+     * It is important to note that this method is called before marking transaction as committing in active txn record. 
+     * So retrieving a transaction from this ordered list is no guarantee that the transaction has been set to committing. 
+     * Similarly, there could be duplicate entries for the same transaction if retried. 
+     * So the ordered ist merely captures the order in which a request may have been received and is only opportunistic 
```

Comment:
I'm wondering if the following situation is possible:

1- We add txn to commit order
2- We get the smallest epoch and find that txn there
3- But, we can't commit it because it is not in `COMMITTING` state yet
4- We remove it from the commit order because we processed its epoch
5- Txn transitions to `COMMITTING`, but now we lost its order information.

How do we avoid such a scenario?

Reply:
no because we dont do `4`. 

We only remove a transaction from committing order once its committed or aborted. 
Other special cases are if the transaction entry in orderer is found to be duplicate. 

The approach is -- 
1. fetch transactions from orderer
2. group transactions by epoch. 
3. within an epoch, fetch transaction status. 
It is important to note that for a writer using sequential txn, if there are multiple transactions in the orderer, all transactions lower than last transaction from the writer cannot be in OPEN state.